Initial mavuika implementation

Charlie-Zheng opened the pull request:
- [x] New character package
- [x] Config in character package
- [x] Run pipeline with added config (generates character curve, talent stats, `.generated.json` files)
- [x] Character key
- [x] Shortcuts for character key
- [x] Update `mode_gcsim.js` with shortcuts for syntax highlighting
- [x] Add Character package to imports
- [x] Normal Attack
- [x] Charge Attack / Aimed Shot
- [x] Skill
- [x] Burst
- [x] A1
- [x] A4
- [x] C1
- [x] C2
- [x] C3
- [x] C4
- [x] C5
- [x] C6
- [x] Other necessary talents (custom dash/jump, low/high plunge, ...)
- [x] Hitlag
- [x] ICD
- [x] StrikeType
- [x] PoiseDMG (blunt attacks only for now)
- [x] Hitboxes
- [x] Attack durability
- [x] Particles
- [x] Frames
- [x] Update documentation
- [x] Xingqiu/Yelan N0 (optional)
- [x] Xianyun Plunge (optional)

Aftermathrar committed:
Update Dehya C2 Skill Mechanics

Removed C2 extending initial cast.
C2 extension corrected to 6s.

Aftermathrar committed:
Update Dehya E proc delay

Changed from 10f to 2f after in-game testing.

Aftermathrar committed:
Update Dehya skill.go

Syntax fix from previous PR?

Aftermathrar committed:
Merge branch 'main' of https://github.com/Aftermathrar/gcsim

Aftermathrar committed:
Update Dehya C2 Field Mechanics

- Add C2 DMG% buff when party member takes damage within field
- Delay adding field by 1f after final Burst hit to avoid self-triggering

Aftermathrar committed:
Update Dehya Field After Burst

- Reapply remaining sanctum DoT ICD after burst

Aftermathrar committed:
Update Dehya C2 Recast Mechanics

- Remove normal attacks resetting the boolean for whether skill had been reused
- Rename boolean for skill reuse to make more sense
- New function to remove field since it is used in several scenarios
- Skill recast function now uses character skill tracking variable for skill duration

Aftermathrar committed:
Merge remote-tracking branch 'upstream/main'

Aftermathrar committed:
Update Dehya Burst Mechanics

- Fixed burst casting skill if starting with skill off cooldown.
- Set burst status to apply on cast to handle actions that can interrupt the burst being used before first hitmark. Will need to evaluate best way to cancel the first queued attack.
- Removed old burstCast tracking variable.

k0l11 committed:
Merge branch 'master' into dehya

k0l11 committed:
fix dehya n1 to n4 and reposition lv.4 multipliers
- remove / 100 as well

k0l11 committed:
fix dehya n1 to n4 hitlag and small frame adjustments

k0l11 committed:
Merge branch 'master' into dehya

k0l11 committed:
Merge branch 'master' into dehya

k0l11 committed:
Merge branch 'master' into dehya

k0l11 committed:
Merge branch 'master' into dehya

k0l11 committed:
more dehya fixes
- implement a1
- implement a4
- implement skill damage mitigation and self DoT
- fix burst frames because no basis on sheet
- fix burst punch/kick state for iframe semantics
- move asc/cons checks into the functions
- move c2 to its own event sub
- rename remove field to pick up field

k0l11 committed:
fix dehya a1 missing trigger icd

k0l11 committed:
refactor dehya c1
- two vars instead of array for flat dmg ratio

k0l11 committed:
refactor dehya c2

k0l11 committed:
refactor dehya c4
- fix c4 icd not being hitlag extended
- adjust c4 heal message
- c4 heal target should always be dehya

k0l11 committed:
refactor dehya c6

k0l11 committed:
fix dehya burst
 - fix slow punch hitmark
 - fix burst interaction with attack/skill during punches
 - add dash -> jump interaction during burst leading to kick
 - refactor jump canceling burst punches
 - min 6 punches idle, max 10 punches <C6, max 15 punches C6 applies now
 - remove burstFrames for constant
 - cannot cancel before first punch so simplified burst func

k0l11 committed:
fix dehya q kick frames

k0l11 committed:
fix dehya q not being hitlag extendable
- doesn't really matter right now because burst punch/kick do not have hitlag

k0l11 committed:
fix dehya e
- fix frames
- fix sac pickup timing to happen right before redeploy instead of on action start
- remove unnecessary hitlag info for non-hitlag attacks
- adjust c2 flag reset to be a bit more future-proof
- fix skill damage DoT having icd (should have no icd)

Charlie-Zheng committed:
Initial mavuika implementation

Aftermathrar committed:
Fix Clorinde eNA AtkSpd Not Applying

Atk Speed did not apply accurately, adjusted animation length to account for attack speed. Hitmarks were moved up slightly (more accurate) to assist with BoL logic in configs.

Aftermathrar committed:
Merge branch 'main' of https://github.com/Aftermathrar/gcsim

Aftermathrar committed:
Initial Charge Attack Logic

The spinning hitbox of Mav's bike is emulated and iterates over enemies for a desired number of frames/hits.
Currently only supports charge[hold=X] and charge[hits=X].

Aftermathrar committed:
Merge branch 'main' into mav-dev

Aftermathrar committed:
Mavuika Charge Attack Logic

The spinning hitbox of Mav's bike is emulated and iterates over enemies for a desired number of frames/hits.
Initial support for charge[hold/hits/final] and charge:X inputs.

Aftermathrar committed:
Merge branch 'mav-dev' of https://github.com/Aftermathrar/gcsim into mav-dev

Aftermathrar committed:
Update charge.go

Updating for golang lint errors.
A lot of commented out code to implement later.

Aftermathrar committed:
Update CA continuation logic

CA inputs now track better between hits and hold durations, keeping the bike positioning and primary target hit icd in memory when inputting multiple consecutive CA commands.

Aftermathrar committed:
Add CA target check and icd

Reworked target list declarations to allow for reliable error checking, returning errors if there are no targets within Mav's CA or the primary is not reachable.

Hit icd is per-enemy and looked up via target key.

Aftermathrar committed:
Update per-enemy hit icd

Created new variable to actually track the implemented per-enemy icd.

Aftermathrar committed:
Add Nightsoul and update CAF queuing

Added duration check for nightsoul status and capped CA duration to it.
CAF now triggers when exiting nightsoul, hitting max CA duration, or attacking from a CA (subject to change).

Aftermathrar committed:
Update charge.go

Added hook logic for new dendro cores spawning during a charge spin.
Adjusted hook start/end to reliably capture events.
Capped hold parameter input to one rotation and fixed empty holds resetting the hit icd for enemies.

Aftermathrar committed:
CA Timing Changes

Updated to allow faster CA start out of dash/NA/Burst.
Moved dash hitmark forward, still a placeholder.
Start of CA now procs Yelan/XQ burst waves.

Aftermathrar committed:
Uptime Timings

Actions out of burst now have detailed timings.
CAF now has a lockout period on the right side of the circle. This mirrors in-game mechanics where the CAF can't start from some angles.

Aftermathrar committed:
Update CAF and skill refresh

Added queuing to CAF to make it cast when using a sequence that doesn't interrupt the CA.
Added skill refresh method for when Mav is on bike and recasts the skill, keeping her on the bike.

Aftermathrar committed:
Update CAF timing

Added initial measured frame timing for CAF.
Improved CAF queuing to handle the CA getting extended with an active hitbox.
Removed ring-state attack queue if exiting bike while outside of NS.

Aftermathrar committed:
Update charge.go

Consolidated queued CAF handling.

Aftermathrar committed:
Skill Frames WIP

Added initial frame counts for most actions.
Added return to c.SkillRecast to fix an error from using it while on cd
Pulled out windup calc to declutter CA method.
Removed NS queue bool, can be handled in CAF when applicable.

Aftermathrar committed:
Add Bike NA Counter Mechanic

Added support for Mavuika's bike counters not resetting on dash.
Added Mavuika's NA not resetting when exiting bike due to NS Blessing depleting.
Switch for getting action.Info is used to distinguish skill method needed, could also forcibly update c.armament to bike upon blessing exit.

Aftermathrar committed:
Update Frames

Updated frames for Skill and Attack actions.

Charlie-Zheng committed:
Merge branch 'gcsim_main' into mav-dev

Aftermathrar committed:
Update NA String Logic

Charlie-Zheng committed:
Merge branch 'mav-dev' of github.com:Aftermathrar/gcsim into mav-dev

Aftermathrar committed:
Add Plunge Attack

Added frames for plunge for bike and non-bike version.
Support for walk-jump into low_plunge.

Aftermathrar committed:
Update charge.go

Comment update on CAF Delay

Aftermathrar committed:
fix lint issues

Only one left, not sure how to solve.

Aftermathrar committed:
Update charge.go

Fix nested statements to cap allowed hold duration to various limits.

Aftermathrar committed:
Fixed CAF Timing

CAF was double-adding partial CA durations when forced into final hit.

Aftermathrar committed:
Update Skill use from Plunge

Recast can queue faster than normal skill use, so this extends the normal skill frames out of plunge to match total frame times. The ratio is slightly skewed, but I think it's the lesser of two evils.
Fixed CAF timing again.

Charlie-Zheng committed:
Merge branch 'mav-dev' of github.com:Aftermathrar/gcsim into mav-dev

Aftermathrar committed:
Add Hitlag Extension to A4

Based on testing, A4 is affected by hitlag.

Charlie-Zheng committed:
Merge branch 'mav-dev' of github.com:Aftermathrar/gcsim into mav-dev

Charlie-Zheng committed:
Made a4 and nightsoul reduce affected by hitlag

jwalk-511 commented on internal/characters/mavuika/cons.go:
```diff
@@ -0,0 +1,205 @@
+		if a.Target.Type() != targets.TargettableEnemy {
+			return
+		}
+		if c.StatusIsActive(c6IcdKey) {
```
Does this status have a function, since the ring can only trigger once every 2 seconds? Or is it just an extra sanity in case the ring hit interval is somehow decreased?

Or does the cb from QueueAttack get called for every target hit, so this is required to prevent multiple triggers from a single ring hit in aoe?

jwalk-511 commented on internal/characters/mavuika/jump.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+func (c *char) Jump(p map[string]int) (action.Info, error) {
+	if !c.StatusIsActive(player.XianyunAirborneBuff) && c.armamentState == bike &&
```
Leaving note for reference for plunge code review. If Mavuika has Xianyun Airborne Buff, Mavuika will do a Xianyun-buffed jump, no matter whether Mav has bike armament state or not.

jwalk-511 commented on internal/characters/mavuika/plunge.go:
```diff
@@ -0,0 +1,213 @@
+func (c *char) LowPlungeAttack(p map[string]int) (action.Info, error) {
+	defer c.Core.Player.SetAirborne(player.Grounded)
+	switch c.Core.Player.Airborne() {
+	case player.AirborneXianyun:
```
Leaving note for plunge reference. If Mav has AirborneXianyun, Mav will perform lowPlungeXY whether or not she has bike state

Aftermathrar replied:
Done, added bike/unbiked versions for XY plunge and unbiked plunge from a walked bike jump.

Aftermathrar committed:
bool name changes

Updated the valid target check to align with what it's actually checking.
Changed bool for final hit to be more intuitive with the conditions triggering it.

Aftermathrar committed:
Update plunge.go

Add support for biked plunges with XY buff.

jwalk-511 commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+// Iterate through CA frames, starting at hitmark
+func (c *char) CalculateValidCollisionFrames(durationCA int, collisionFrames [2]int, lastHitFrame int) []int {
+	validFrames := []int{}
+	currentFrame := 35 // Spin starts on frame 35 of CA anim (full windup)
```
I don't think this function works when called from the ca hook cb. If a dendro core is created on some arbitrary frame in the middle of a function, Time Since Start should be something like c.Core.F - c.caState.srcF (but accounting for hitlag).

Aftermathrar replied:
True, I changed this recently from the calc you had to cAtkFrames (when trying to separate out windup) and didn't realize the implication. I will change it back.

jwalk-511 commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+	// Do not allow starting with a partial CA hold
+	if durationCA > 0 && c.caState.cAtkFrames > 0 {
+		// Cap duration to lowest of 1 spin, remaining NS, or max CA time
+		durationCA = min(durationCA, bikeChargeAttackSpinFrames, nightSoulDuration, bikeChargeAttackMaximumDuration-c.caState.cAtkFrames)
```
Recommend adding to documentation that the "hold" parameter represents time to remain in CA state not counting windup time, as well as the fact that the minimum is 1f and maximum is 1 complete rotation (bikeChargeAttackSpinFrames). Also note that by default, the logic is not to hold for 0f but to spin until the primary target is hit.

Also recommend leaving a note for the skill itself (not for the param), that the duration will automatically be cut short if max CA time is reached or NS expires- but this note is less important since it's a reasonable assumption to make anyway.

Aftermathrar replied:
Updated documentation to add this.

jwalk-511 reviewed (commented):
Review is not done, just leaving comments so far as a starting point (Notably, I have reviewed all functions called by BikeCharge from line 121 to 147, and functions called by those functions, recursively. But I have not reviewed functions only called by BikeCharge after line 147, including `HoldBikeChargeAttack` and `CountBikeChargeAttack`)

imring committed:
bugfixes pt. 1

Aftermathrar commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+		if g == nil {
+			continue
+		}
+		switch g.GadgetTyp() {
```
I might be interpreting this wrong, but it looks like all gadgets are targettable, so a default case would be handling pretty much every other gadget type and may not know if something is "new" unless the current case select has every gadget type explicitly handled. I think either way would require a bug report in the future, but I might be missing an easier solution here.

jwalk-511 replied:
I think you're right; I was assuming that most gadgets would have some flag indicating they shouldn't be targeted by player abilities but if that isn't present then it isn't a super huge issue.

It could only possibly matter if some future character can create a gadget that can be hit by any other character, which is a rare case anyway

Charlie-Zheng replied:
Generally if a player cannot hit the gadget, there is no reason to actually code the object as a gadget.

Aftermathrar commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+
+func (c *char) buildValidTargetList() []HittableEntity {
+	enemies := c.Core.Combat.EnemiesWithinArea(combat.NewCircleHitOnTarget(c.Core.Combat.Player(), nil, 8), nil)
+	hittableEnemies := []HittableEntity{}
```
Done.

Aftermathrar commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+	case *geometry.Circle:
+		enemyRadius = v.Radius() // Rt
+	default:
+		return false
```
Added error saying Mav needs circular hitboxes for CA calcs.

Aftermathrar commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+			targetIndex := int(t.Entity.Key())
+			c.caState.LastHit[targetIndex] = 0
+		}
+		c.Core.Tasks.Add(func() {
+			c.bikeChargeAttackHook()
+		}, bikeChargeAttackStartupHitmark-1)
```
Went with the first option, since cores could be missed in the scenario you described.

Aftermathrar commented on internal/characters/mavuika/charge.go:
```diff
@@ -0,0 +1,763 @@
+	return true, action.Info{}, nil
+}
+
+// Important Events: OnTargetDied, OnTargetMoved (also emits on player move?), OnDendroCore
```
Comment updated.

Aftermathrar committed:
Update charge, dash, docs

Updated docs for Mavuika's actions and params.
Added bike dash handling for specific frame counts.
Added an error when a target has a hitbox shape that does not work with the CA calculation.
Changed hook timing to capture dendro cores that spawn during windup.
Updated comments.
Fixed CAF not meeting the minimum CA duration if used as the first action.
Added 'buffered' param to allow for delaying CA after burst in order to proc N0 triggers.

Aftermathrar committed:
Fixing lint issues

lint bugfixes

Aftermathrar committed:
Update character_data.json

Add note about approximated aoe

Aftermathrar committed:
Update docs

Added hitlag and aoe information.

imring committed:
fix docs

imring committed:
pipeline
